src/wasm/wasm-binary.cpp

```diff
@@ -950,6 +952,99 @@ void WasmBinaryWriter::finishUp() {
   }
 }
 
+void WasmBinaryWriter::writeType(Type type) {
+  if (type.isRef()) {
+    auto heapType = type.getHeapType();
+    if (heapType.isSignature()) {
+      if (type.isNullable()) {
+        o << S32LEB(BinaryConsts::EncodedType::nullable);
+      } else {
+        o << S32LEB(BinaryConsts::EncodedType::nonnullable);
+      }
+      writeHeapType(heapType);
```

Comment:
We don't call `writeHeapType` when `heapType` is not a signature?

Reply:
I think maybe we should for GC. But for the typed function references proposal, all we have are signatures.

Maybe there's no harm in generalizing it here though, I'll update with that. Seems like it can't hurt, may save time later, and is shorter anyhow.

Reply:
I had to revert that - it turns out we already have a testcase in the test suite that breaks because of the change, presumably because we need to fix the other side (reading and writing) as well. Let's leave GC for later then.

Reply:
That sounds fine; it'd be better to leave a TODO here maybe, because when we currently call it only for signatures.